direct_access.cpp

```diff
@@ -0,0 +1,133 @@
+
+  // get direct access pointers
+  moab::EntityHandle* conn_ptr;
+  moab::EntityHandle first_element = all_tris.front();
+  int stride, n_tris;
+  rval = MBI->connect_iterate(all_tris.begin(), all_tris.end(), conn_ptr, stride, n_tris);
+  MB_CHK_SET_ERR(rval, "Failed to get connectivity pointer");
+
+  double* x_vals;
+  double* y_vals;
+  double* z_vals;
+
+  moab::Range all_verts;
+  rval = MBI->get_entities_by_dimension(0, 0, all_verts);
+  MB_CHK_SET_ERR(rval, "Failed to get all vertices in the model");
+
+  int n_vertices;
+  rval = MBI->coords_iterate(all_verts.begin(), all_verts.end(), x_vals, y_vals, z_vals, n_vertices);
+  MB_CHK_SET_ERR(rval, "Failed to get coordinate pointers");
+
+  size_t n_samples = 10000000;
+
+  // write binary file if indicated by "-w" flag
+  bool write_vals = argc > 1 && std::string(argv[1]) == "-w";
+
+  Timer t;
+  std::cout << "Running " << n_samples << " samples." << std::endl;
+
+  // vector holding a random set of indices to use as samples
+  std::vector<size_t> random_indices(n_samples);
+  srand(100); // ensure random number sequence is the same
+  for(int i = 0; i < n_samples; i++) {
+    random_indices.at(i) = rand() % all_tris.size();
+  }
+
+  std::vector<std::string> results(n_samples);
+
+  t.start();
+  #pragma omp parallel for shared(MBI, write_vals, random_indices, results)
+  for(size_t i = 0; i < n_samples; i++) {
+
+    // get random index for this sample
+    size_t idx = random_indices[i];
+
+    // query the triangle connectivity
+    moab::EntityHandle eh = all_tris[idx];
+
+    moab::Range conn;
+    size_t offset = stride * (eh - all_tris.front());
+
+    moab::EntityHandle v0 = conn_ptr[offset];
+    moab::EntityHandle v1 = conn_ptr[offset + 1];
+    moab::EntityHandle v2 = conn_ptr[offset + 2];
+
+    // MOAB's EntityHandle generation starts at one, but C++
+    // arrays start at zero
+    // handle off-by-one when determining the coordinates
+    double coords[3][3];
+    coords[0][0] = x_vals[v0 - 1];
+    coords[0][1] = y_vals[v0 - 1];
+    coords[0][2] = z_vals[v0 - 1];
+
+    coords[1][0] = x_vals[v1 - 1];
+    coords[1][1] = y_vals[v1 - 1];
+    coords[1][2] = z_vals[v1 - 1];
+
+    coords[2][0] = x_vals[v2 - 1];
+    coords[2][1] = y_vals[v2 - 1];
+    coords[2][2] = z_vals[v2 - 1];
+
+    // write values if requested
+    if (write_vals) {
+      std::stringstream ss;
+
+      // the entity we're querying
+      ss.write((const char*)&eh, sizeof(moab::EntityHandle));
+
+      // the connectivity of that entity
+      ss.write((const char*)&v0, sizeof(moab::EntityHandle));
+      ss.write((const char*)&v1, sizeof(moab::EntityHandle));
+      ss.write((const char*)&v2, sizeof(moab::EntityHandle));
+
+      for (int i = 0; i < 3; i++) {
+        for (int j = 0; j < 3; j++) {
+          ss.write((const char*)&coords[i][j], sizeof(double));
+        }
+      }
+
+      results[i] = ss.str();
+    }
+  }
+  t.stop();
```

Comment:
The question here is whether your "kernel" is representative. IMO, the kernel here, which just streams coordinates into a string, isn't representative. It's hard to guess whether it's more expensive than need be (because it uses a vector of strings, which involves lots of small allocations) or less (because it doesn't really do any floating point math).

Instead, I recommend you do something typical, like compute the triangle centroid or triangle normal.

Reply:
These stream operations only occur if the `-w` flag is passed to the program. That option is intended to check that both methods access the same triangles, coordinates, etc. rather than to measure performance. 

That aside, this test has more to do with general access performance through rather than performance directly related to ray tracing/Monte Carlo. I don't know how important having a representative kernel is.